### PR TITLE
[AutoParallel] selective recompute

### PR DESCRIPTION
**Pull request by zhaoyinglia**
### PR types

New features
### PR changes

Others
### Describe

- auto_parallel support selective recompute
- two methods to use this feature:
```
# method 1:
from paddle.distributed.fleet import auto
recompute = auto.Strategy().recompute
recompute.enable = True
recompute.no_recompute_segments = [0]

auto.recompute(func)

# method 2:
from paddle.distributed.fleet import auto
recompute = auto.Strategy().recompute
recompute.enable = True
recompute.checkpoints = ['tmp1', 'tmp3']
recompute.no_recompute_segments = [0]
```

### CODE REVIEW

**Review by XieYunshen — APPROVED** [review]
LGTM for 
`set_tests_properties(test_selective_recompute PROPERTIES TIMEOUT 50)`

**Review by aoyulong — APPROVED** [review]
LGTM

File: python/paddle/distributed/auto_parallel/interface.py [hunk 1]
```diff
@@ -209,7 +215,9 @@ def __call__(self, *args, **kwargs):
 
             for idx in range(op_size, new_op_size):
                 op = cur_block.ops[idx]
-                op._set_attr("is_recompute@auto_parallel", True)
+                op._set_attr(
```
Comment (JZ-LIANG) [inline python/paddle/distributed/auto_parallel/interface.py:218]: Use another attr like dist attr for this information.
op_namescope is supposed to be used by many other logic. 
if the modified op is not under fully control by auto parallel(like the allreduce only added and used by auto parallel logic),  you might remove the information of  other.

Comment (zhaoyinglia) [reply python/paddle/distributed/auto_parallel/interface.py:218]: Ok, the attribute will be moved to dist_attr in the next PR.

**Pull request merged**